h2o-algos/src/main/java/hex/tree/TreeHandler.java

```diff
@@ -175,6 +280,116 @@ private static void append(final int[] rightChildren, final int[] leftChildren,
                 discoveredNodes, pointer, true);
     }
 
+    private static ArrayList<Integer> extractInternalIds(TreeProperties properties) {
+        int nodeId = 0;
+        ArrayList<Integer> nodeIds = new ArrayList<>();
+        nodeIds.add(nodeId);
+        for (int i = 0; i < properties._leftChildren.length; i++) {
+            if (properties._leftChildren[i] != -1) {
+                nodeId++;
+                nodeIds.add(properties._leftChildren[i]);
+                properties._leftChildren[i] = nodeId;
+            }
+
+            if (properties._rightChildren[i] != -1) {
+                nodeId++;
+                nodeIds.add(properties._rightChildren[i]);
+                properties._rightChildren[i] = nodeId;
+            }
+        }
+        return nodeIds;
+    }
+
+    private static void fillLanguagePathRepresentation(TreeProperties properties, String[] domainValues) {
+        ArrayList<Integer> nodeIds = extractInternalIds(properties);
+        nodeIds.forEach((list_path_id) -> {
+            int index = nodeIds.indexOf(list_path_id);
+            properties._languagePathsRepresentations[index] = fillNodePath(list_path_id, nodeIds, false, properties, domainValues);
+        });
+    }
+
+    private static String fillNodePath(int nodeId, ArrayList<Integer> nodeIds, boolean valuePrinted, TreeProperties properties, String[] domainValues){
+        int parentIndex = -1;
+        int parentId = -1;
+        String condition = "";
+        String nodePathr = "";
+        int currentNodeIndex = nodeIds.indexOf(nodeId);
+        if (!valuePrinted) {
+            // print prediction value
+            nodePathr += "Predicted value: " + properties._predictions[currentNodeIndex] + "\n";
+            valuePrinted = true;
+            nodePathr += fillNodePath(nodeId, nodeIds, valuePrinted, properties, domainValues);
+        } else {
+            // print conditions leading to prediction value
+            int[] leftChildren = properties._leftChildren;
+            int[] rightChildren = properties._rightChildren;
+            if (IntStream.of(leftChildren).anyMatch(i -> i == currentNodeIndex)) {
```

Comment:
How efficient is `anyMatch` on IntStream ? Looks like it translates to `ForEachOp`. At least binary search would be nice - the array `lefChildren` should be sorted in ascending order, there should NOT be smaller numbers at lower index than there are at higher index. We have it implemented as `Array.binarySearch`.

Reply:
fixed

Reply:
binary search can't be used - there are cases when properties._leftChildren is like ... 62 64 66 -1 68 70 72 ... 86 88 90 -1 -1 -1 ... -> properties._leftChildrenNormalized are created like ...31 33 35 -1 37 39 ... 55 57 59 -1 -1 -1 ... hence even when I cut -1's from right it is possible that leftChildren/_leftChildrenNormalized  is not sorted and I can't sort it because it would invalidate information... (this is the case from pyunit_tree.py which is failing after this change)

Reply:
I see. Let's have it this way then and optimize later.